django/contrib/admin/templates/admin/change_form.html

```diff
@@ -34,7 +40,8 @@
 {% endif %}
 {% endblock %}
 <form {% if has_file_field %}enctype="multipart/form-data" {% endif %}{% if form_url %}action="{{ form_url }}" {% endif %}method="post" id="{{ opts.model_name }}_form" novalidate>{% csrf_token %}{% block form_top %}{% endblock %}
-<div>
+<div id="changeform">
+{% if action_form %}{% admin_actions %}{% endif %}
```

Comment:
Have you considered including a hidden input here, something like:
```html
<input type="hidden" name="{{ action_checkbox_name }}" value="{{ obj.pk|unlocalize }}">
```

The idea being that
```python
selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
```
returns a list of one item being the object PK for the change form.

Then you can still handle everything as previously rather than now having to handle the single option case? I'm not sure if this will work but I think it's worth a try? What do you think?

Reply:
> ```
> action_checkbox_name
> ```

Thank you Sarah! I've tried that option but I couldn't make it work, and also I had to send via the context the `ACTION_CHECKBOX_NAME` and the `obj`

I finally solved the issue adding this line: https://github.com/django/django/pull/16012/files#diff-e77dfa6935de3fac94356ffd1029a679f8aa23f2979fe9b31aea7f93e1dccad3R1887

With that `selected` will be always a `queryset` in case of the detail view: a QuerySet of 1 object. IMO easier to handle when you define your custom action. And no backwards compatibility breaks

Let me know what do you think!

Reply:
I had a bit of a further look into it and sorry there was some mistakes in the suggestion but I can get this to work with the following
```diff
diff --git a/django/contrib/admin/actions.py b/django/contrib/admin/actions.py
index 805ae0feca..314fe1becb 100644
--- a/django/contrib/admin/actions.py
+++ b/django/contrib/admin/actions.py
@@ -31,11 +31,6 @@ def delete_selected(modeladmin, request, queryset):
     opts = modeladmin.model._meta
     app_label = opts.app_label
 
-    # Handle the case when delete action is called from change view and
-    # it is a single object
-    if not isinstance(queryset, QuerySet):
-        queryset = modeladmin.model.objects.filter(id=queryset.id)
-
     # Populate deletable_objects, a data structure of all related objects that
     # will also be deleted.
     (
diff --git a/django/contrib/admin/options.py b/django/contrib/admin/options.py
index 9961ddd729..ceb18a6609 100644
--- a/django/contrib/admin/options.py
+++ b/django/contrib/admin/options.py
@@ -1615,7 +1615,7 @@ class ModelAdmin(BaseModelAdmin):
         """
         return self._response_post_save(request, obj)
 
-    def response_action(self, request, queryset, change):
+    def response_action(self, request, queryset, change=False):
         """
         Handle an admin action. This is called if a request is POSTed to the
         changelist or changeform; it returns an HttpResponse if the action was
@@ -1652,26 +1652,25 @@ class ModelAdmin(BaseModelAdmin):
         # If the form's valid we can handle the action.
         if action_form.is_valid():
             action = action_form.cleaned_data["action"]
+            select_across = action_form.cleaned_data["select_across"]
             func = self.get_actions(request)[action][0]
-            if not change:
-                select_across = action_form.cleaned_data["select_across"]
-                # Get the list of selected PKs. If nothing's selected, we can't
-                # perform an action on it, so bail. Except we want to perform
-                # the action explicitly on all objects.
-                selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
-                if not selected and not select_across:
-                    # Reminder that something needs to be selected or nothing will
-                    # happen
-                    msg = _(
-                        "Items must be selected in order to perform "
-                        "actions on them. No items have been changed."
-                    )
-                    self.message_user(request, msg, messages.WARNING)
-                    return None
 
-                if not select_across:
-                    # Perform the action only on the selected objects
-                    queryset = queryset.filter(pk__in=selected)
+            # Get the list of selected PKs. If nothing's selected, we can't
+            # perform an action on it, so bail. Except we want to perform
+            # the action explicitly on all objects.
+            selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
+            if not selected and not select_across:
+                # Reminder that something needs to be selected or nothing will happen
+                msg = _(
+                    "Items must be selected in order to perform "
+                    "actions on them. No items have been changed."
+                )
+                self.message_user(request, msg, messages.WARNING)
+                return None
+
+            if not select_across:
+                # Perform the action only on the selected objects
+                queryset = queryset.filter(pk__in=selected)
 
             response = func(self, request, queryset)
 
(main-3.12) sarahboyce@sarahboyce-ThinkPad-E15-Gen-4:~/Documents/django/tests$ git diff
diff --git a/django/contrib/admin/actions.py b/django/contrib/admin/actions.py
index 805ae0feca..314fe1becb 100644
--- a/django/contrib/admin/actions.py
+++ b/django/contrib/admin/actions.py
@@ -31,11 +31,6 @@ def delete_selected(modeladmin, request, queryset):
     opts = modeladmin.model._meta
     app_label = opts.app_label
 
-    # Handle the case when delete action is called from change view and
-    # it is a single object
-    if not isinstance(queryset, QuerySet):
-        queryset = modeladmin.model.objects.filter(id=queryset.id)
-
     # Populate deletable_objects, a data structure of all related objects that
     # will also be deleted.
     (
diff --git a/django/contrib/admin/options.py b/django/contrib/admin/options.py
index 9961ddd729..ceb18a6609 100644
--- a/django/contrib/admin/options.py
+++ b/django/contrib/admin/options.py
@@ -1615,7 +1615,7 @@ class ModelAdmin(BaseModelAdmin):
         """
         return self._response_post_save(request, obj)
 
-    def response_action(self, request, queryset, change):
+    def response_action(self, request, queryset, change=False):
         """
         Handle an admin action. This is called if a request is POSTed to the
         changelist or changeform; it returns an HttpResponse if the action was
@@ -1652,26 +1652,25 @@ class ModelAdmin(BaseModelAdmin):
         # If the form's valid we can handle the action.
         if action_form.is_valid():
             action = action_form.cleaned_data["action"]
+            select_across = action_form.cleaned_data["select_across"]
             func = self.get_actions(request)[action][0]
-            if not change:
-                select_across = action_form.cleaned_data["select_across"]
-                # Get the list of selected PKs. If nothing's selected, we can't
-                # perform an action on it, so bail. Except we want to perform
-                # the action explicitly on all objects.
-                selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
-                if not selected and not select_across:
-                    # Reminder that something needs to be selected or nothing will
-                    # happen
-                    msg = _(
-                        "Items must be selected in order to perform "
-                        "actions on them. No items have been changed."
-                    )
-                    self.message_user(request, msg, messages.WARNING)
-                    return None
 
-                if not select_across:
-                    # Perform the action only on the selected objects
-                    queryset = queryset.filter(pk__in=selected)
+            # Get the list of selected PKs. If nothing's selected, we can't
+            # perform an action on it, so bail. Except we want to perform
+            # the action explicitly on all objects.
+            selected = request.POST.getlist(helpers.ACTION_CHECKBOX_NAME)
+            if not selected and not select_across:
+                # Reminder that something needs to be selected or nothing will happen
+                msg = _(
+                    "Items must be selected in order to perform "
+                    "actions on them. No items have been changed."
+                )
+                self.message_user(request, msg, messages.WARNING)
+                return None
+
+            if not select_across:
+                # Perform the action only on the selected objects
+                queryset = queryset.filter(pk__in=selected)
 
             response = func(self, request, queryset)
 
@@ -1881,17 +1880,14 @@ class ModelAdmin(BaseModelAdmin):
         actions = self.get_actions(request)
         if request.method == "POST":
             if actions and request.POST.get("action", ""):
-                action_failed = False
                 response = self.response_action(
                     request,
-                    self.get_queryset(request).filter(id=object_id),
-                    change=not add,
+                    self.get_queryset(request),
+                    change=True
                 )
                 if response:
                     return response
                 else:
-                    action_failed = True
-                if action_failed:
                     # Redirect back to the changelist page to avoid resubmitting the
                     # form if the user refreshes the browser or uses the "No, take
                     # me back" button on the action confirmation page.
@@ -1968,7 +1964,7 @@ class ModelAdmin(BaseModelAdmin):
 
         # Build the action form and populate it with available actions.
         if actions and not add:
-            action_form = self.action_form(auto_id=None)
+            action_form = self.action_form(initial={"change": True}, auto_id=None)
             action_form.fields["action"].choices = self.get_action_choices(
                 request, change=True
             )
@@ -1987,6 +1983,7 @@ class ModelAdmin(BaseModelAdmin):
             "to_field": to_field,
             "media": media,
             "action_form": action_form,
+            "action_checkbox_name": helpers.ACTION_CHECKBOX_NAME,
             "inline_admin_formsets": inline_formsets,
             "errors": helpers.AdminErrorList(form, formsets),
             "preserved_filters": self.get_preserved_filters(request),
@@ -2085,7 +2082,7 @@ class ModelAdmin(BaseModelAdmin):
         ):
             if selected:
                 response = self.response_action(
-                    request, queryset=cl.get_queryset(request), change=False
+                    request, queryset=cl.get_queryset(request)
                 )
                 if response:
                     return response
@@ -2109,7 +2106,7 @@ class ModelAdmin(BaseModelAdmin):
         ):
             if selected:
                 response = self.response_action(
-                    request, queryset=cl.get_queryset(request), change=False
+                    request, queryset=cl.get_queryset(request)
                 )
                 if response:
                     return response
diff --git a/django/contrib/admin/templates/admin/change_form.html b/django/contrib/admin/templates/admin/change_form.html
index 43dc9fef3c..c07a166fe5 100644
--- a/django/contrib/admin/templates/admin/change_form.html
+++ b/django/contrib/admin/templates/admin/change_form.html
@@ -1,5 +1,5 @@
 {% extends "admin/base_site.html" %}
-{% load i18n admin_urls static admin_modify %}
+{% load i18n l10n admin_urls static admin_modify %}
 
 {% block extrahead %}{{ block.super }}
 <script src="{% url 'admin:jsi18n' %}"></script>
@@ -41,7 +41,7 @@
 {% endblock %}
 <form {% if has_file_field %}enctype="multipart/form-data" {% endif %}{% if form_url %}action="{{ form_url }}" {% endif %}method="post" id="{{ opts.model_name }}_form" novalidate>{% csrf_token %}{% block form_top %}{% endblock %}
 <div id="changeform">
-{% if action_form %}{% admin_actions %}{% endif %}
+{% if action_form %}{% admin_actions %}<form method="post">{% csrf_token %}<input type="hidden" name="{{ action_checkbox_name }}" value="{{ object_id|unlocalize }}"></form>{% endif %}
 {% if is_popup %}<input type="hidden" name="{{ is_popup_var }}" value="1">{% endif %}
 {% if to_field %}<input type="hidden" name="{{ to_field_var }}" value="{{ to_field }}">{% endif %}
 {% if save_on_top %}{% block submit_buttons_top %}{% submit_row %}{% endblock %}{% endif %}
diff --git a/tests/admin_views/test_actions.py b/tests/admin_views/test_actions.py
index 77a67b0e6e..c8c5f8c51c 100644
--- a/tests/admin_views/test_actions.py
+++ b/tests/admin_views/test_actions.py
@@ -645,6 +645,7 @@ class AdminDetailActionsTest(TestCase):
                 "email": "foo@bar.com",
                 "_save": "Save",
                 "action": "",
+                ACTION_CHECKBOX_NAME: [self.s1.pk],
             },
         )
         external_subscriber = ExternalSubscriber.objects.get()
@@ -663,6 +664,7 @@ class AdminDetailActionsTest(TestCase):
                 "email": "foo@bar.com",
                 "_save": "Save",
                 "action": "external_mail",
+                ACTION_CHECKBOX_NAME: [self.s1.pk],
             },
         )
         self.assertEqual(len(mail.outbox), 1)
@@ -679,7 +681,7 @@ class AdminDetailActionsTest(TestCase):
 
         response = self.client.post(
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
-            {"action": "download"},
+            {"action": "download", ACTION_CHECKBOX_NAME: [self.s1.pk]},
         )
         content = b"".join(list(response))
         self.assertEqual(
@@ -696,7 +698,7 @@ class AdminDetailActionsTest(TestCase):
         self.user.user_permissions.add(permission)
         response = self.client.post(
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
-            {"action": "delete_selected"},
+            {"action": "delete_selected", ACTION_CHECKBOX_NAME: [self.s1.pk]},
         )
         self.assertEqual(response.status_code, 200)
         self.assertContains(
@@ -708,6 +710,7 @@ class AdminDetailActionsTest(TestCase):
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
             {
                 "action": "delete_selected",
+                ACTION_CHECKBOX_NAME: [self.s1.pk],
                 "post": "yes",
             },
         )
@@ -718,7 +721,7 @@ class AdminDetailActionsTest(TestCase):
         # User doesn't have the permission to run the custom action.
         response = self.client.post(
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
-            {"action": "custom_action"},
+            {"action": "custom_action", ACTION_CHECKBOX_NAME: [self.s1.pk]},
         )
         self.assertEqual(response.status_code, 302)
 
@@ -732,7 +735,7 @@ class AdminDetailActionsTest(TestCase):
         self.user.user_permissions.add(permission)
         response = self.client.post(
             reverse("admin:admin_views_externalsubscriber_change", args=[self.s1.pk]),
-            {"action": "custom_action"},
+            {"action": "custom_action", ACTION_CHECKBOX_NAME: [self.s1.pk]},
         )
         self.assertEqual(response.status_code, 200)
         self.assertEqual(response.content, b"OK")
diff --git a/tests/admin_views/tests.py b/tests/admin_views/tests.py
index e0a4926b91..82917bd4dc 100644
--- a/tests/admin_views/tests.py
+++ b/tests/admin_views/tests.py
@@ -4841,6 +4841,7 @@ class AdminInheritedInlinesTest(TestCase):
         )
         names = name_re.findall(response.content)
         names.remove(b"csrfmiddlewaretoken")
+        names.remove(b"csrfmiddlewaretoken")
         # make sure we have no duplicate HTML names
         self.assertEqual(len(names), len(set(names)))
```